pkg/controller/controller_crd_event_handler.go

```diff
@@ -25,29 +25,21 @@ type watchState struct {
 type crdEventHandler struct {
 	ctx context.Context
 	*BundleController
-	mx       sync.Mutex            // protects the map
 	watchers map[string]watchState // CRD name -> state
 }
 
 func (h *crdEventHandler) OnAdd(obj interface{}) {
 	crd := obj.(*apiext_v1b1.CustomResourceDefinition)
-	if crd.Name == smith.BundleResourceName {
-		return
+	if h.ensureWatch(crd) {
+		h.rebuildBundles(crd, "added")
 	}
-	func() {
-		h.mx.Lock()
-		defer h.mx.Unlock()
-		h.watch(crd)
-	}()
-	h.rebuildBundles(crd, "added")
 }
 
 func (h *crdEventHandler) OnUpdate(oldObj, newObj interface{}) {
 	newCrd := newObj.(*apiext_v1b1.CustomResourceDefinition)
-	if newCrd.Name == smith.BundleResourceName {
-		return
+	if h.ensureWatch(newCrd) {
```

Comment:
Yes.
Add event is received and it only informs us that a CRD object has been added. It may not have been processed by the CRD controller yet - we can tell it by looking at the conditions. Then, once controller has processed the CRD, it updates it with proper conditions and we get an update event.

In case when CRD existed before Smith was started, add event may deliver an object that has the conditions set properly. So we need to check them in both OnAdd and OnUpdate. I've added comments.
Does it make sense?